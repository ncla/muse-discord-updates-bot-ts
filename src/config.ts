import dotenv from 'dotenv'

dotenv.config()

// console.log(process.env)

export interface IConfig {
    app: {
        debug: boolean,
    }
    services: {
        crtsh: {
            domain: string | undefined,
        },
        musebootlegs: {
            username: string | undefined,
            password: string | undefined,
            user_agent: string | undefined,
        },
        youtube: {
            uploads_api_key: string | undefined,
            playlists_api_key: string | undefined,
        }
    }
    fetchables: {
        youtube: {
            username: string,
            channel_id: string,
            uploads_playlist_id: string,
            author_image_url: string,
            uploads: boolean,
            playlists: boolean,
        }[]
    }
    webhooks: {
        discord: {
            id: string | undefined,
            token: string | undefined,
        }
    },
    pingable_mention_ids: {
        discord: {
            musebootlegs_torrents: string | undefined,
            youtube_uploads: string | undefined,
            youtube_playlist_videos: string | undefined,
            domain_certificates: string | undefined,
            gigs: string | undefined,
            store: string | undefined,
            musewiki: string | undefined,
        }
    }
}

const config: IConfig = {
    app: {
        debug: process.env.APP_DEBUG === 'true',
    },
    services: {
        crtsh: {
            domain: process.env.CRTSH_DOMAIN,
        },
        musebootlegs: {
            username: process.env.MUSEBOOTLEGS_USERNAME,
            password: process.env.MUSEBOOTLEGS_PASSWORD,
            user_agent: process.env.MUSEBOOTLEGS_USER_AGENT,
        },
        youtube: {
            uploads_api_key: process.env.YOUTUBE_UPLOADS_API_KEY,
            playlists_api_key: process.env.YOUTUBE_PLAYLISTS_API_KEY,
        }
    },
    fetchables: {
        youtube: [
            {
                'username': 'muse',
                'channel_id': 'UCGGhM6XCSJFQ6DTRffnKRIw',
                'uploads_playlist_id': 'UUGGhM6XCSJFQ6DTRffnKRIw',
                'author_image_url': 'https://yt3.ggpht.com/a/AATXAJyCalVyFs_seZmv6CVDdMB5iyI_5L2c1_OBJA=s88-c-k-c0xffffffff-no-rj-mo',
                'uploads': process.env.YOUTUBE_MUSE_UPLOADS === 'true',
                'playlists': process.env.YOUTUBE_MUSE_PLAYLISTS === 'true'
            }
        ]
    },
    webhooks: {
        discord: {
            id: process.env.DISCORD_WEBHOOK_ID,
            token: process.env.DISCORD_WEBHOOK_TOKEN,
        }
    },
    pingable_mention_ids: {
        discord: {
            musebootlegs_torrents: process.env.DISCORD_ROLE_ID_MUSE_BOOTLEGS,
            youtube_uploads: process.env.DISCORD_ROLE_ID_YOUTUBE_UPLOADS,
            youtube_playlist_videos: process.env.DISCORD_ROLE_ID_YOUTUBE_PLAYLISTS,
            domain_certificates: process.env.DISCORD_ROLE_ID_DOMAIN_CERTIFICATES,
            gigs: process.env.DISCORD_ROLE_ID_GIGS,
            store: process.env.DISCORD_ROLE_ID_STORE,
            musewiki: process.env.DISCORD_ROLE_ID_MUSEWIKI,
        }
    }
};

export default config;