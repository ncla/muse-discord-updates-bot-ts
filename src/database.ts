import SQLite from 'better-sqlite3'
import { Kysely, SqliteDialect, LogEvent } from 'kysely'
import config from "../src/config";
import {
    ColumnType,
    Generated,
    Insertable,
    JSONColumnType,
    Selectable,
    Updateable,
} from 'kysely'
import {Update} from "@/src/updates";

const dialect = new SqliteDialect({
    database: new SQLite('./data/db.sqlite3'),
})

export function queryLogging(event: LogEvent) {
    if (event.level === "error") {
        console.error("Query failed: ", {
            durationMs: event.queryDurationMillis,
            error: event.error,
            sql: event.query.sql,
            params: event.query.parameters,
        });
    } else {
        console.log("Query executed: ", {
            durationMs: event.queryDurationMillis,
            sql: event.query.sql,
            params: event.query.parameters,
        });
    }
}

export const db = new Kysely<Database>({
    dialect,
    log: config.app.debug ? queryLogging : undefined
})

export interface Database {
    updates: UpdatesTable,
    youtube_playlists: YoutubePlaylistsTable
}

export interface YoutubePlaylistsTable {
    id: Generated<number>

    playlist_id: ColumnType<string, string, never>

    video_count: ColumnType<number, number, number>

    created_at: ColumnType<Date, string | undefined, never>
}

export type ReturnableYoutubePlaylistRecord = Selectable<YoutubePlaylistsTable>
export type InsertableYoutubePlaylistRecord = Insertable<YoutubePlaylistsTable>
export type UpdateableRecordYoutubePlaylist = Updateable<YoutubePlaylistsTable>

export interface UpdatesTable {
    id: Generated<number>

    type: ColumnType<string, string, never>

    unique_id: ColumnType<string, string, never>

    // TODO: Is there something better than simply specifying string type for JSON serialized data?
    data: JSONColumnType<Update | null, Update | string | undefined, Update | string | undefined>

    created_at: ColumnType<Date, string | undefined, never>
}

export type SelectableUpdateRecord = Selectable<UpdatesTable>
export type InsertableUpdateRecord = Insertable<UpdatesTable>
export type UpdateableRecordUpdate = Updateable<UpdatesTable>
