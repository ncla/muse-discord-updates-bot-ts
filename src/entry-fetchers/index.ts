import {Update} from "@/src/updates";

export interface EntryFetcher {
    fetch(): Promise<Update[]>;
}