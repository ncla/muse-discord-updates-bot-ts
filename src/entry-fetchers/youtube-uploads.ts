import {EntryFetcher} from "./index";
import config, {IConfig} from "../config";
// import * as util from "node:util";
import {createBlankUnprocessedUpdate, UnprocessedUpdateEntry, UpdateType} from "../update";

export class YoutubeUploads implements EntryFetcher
{
    async fetch()
    {
        // console.log(util.inspect(config, { depth: 5 }));

        const apiKey = config.services.youtube.uploads_api_key

        if (apiKey === undefined) {
            throw new Error('Youtube uploads API key is not set')
        }

        let entries: UnprocessedUpdateEntry[] = []

        for (const channel of config.fetchables.youtube) {
            let url = new URL(`https://www.googleapis.com/youtube/v3/playlistItems`);

            url.searchParams.append('playlistId', channel.uploads_playlist_id)
            url.searchParams.append('maxResults', '25')
            url.searchParams.append('part', 'snippet')
            url.searchParams.append('key', apiKey)

            const response = await fetch(url.toString());

            if (!response.ok) {
                throw new Error(`Response status: ${response.status}`);
            }

            let json: GoogleApiYouTubePaginationInfo<GoogleApiYouTubePlaylistItemResource> = await response.json();

            const channelUploadEntries: UnprocessedUpdateEntry[] = json.items.map((item): UnprocessedUpdateEntry => {
                return this.mapPlayListItemToEntry(item, channel);
            });

            entries = [...entries, ...channelUploadEntries]
        }

        return entries;
    }

    // TODO: channel typehint seems quite heavy. type-hint without nested properties instead?
    private mapPlayListItemToEntry(
        item: GoogleApiYouTubePlaylistItemResource,
        channel: IConfig['fetchables']['youtube'][number]
    ): UnprocessedUpdateEntry
    {
        return {
            ...createBlankUnprocessedUpdate(),
            uniqueId: `${channel.channel_id}_${item.snippet.resourceId.videoId}`,
            type: UpdateType.YOUTUBE_UPLOAD,
            id: item.snippet.resourceId.videoId,
            title: item.snippet.title,
            content: item.snippet.description ? item.snippet.description : null,
            url: `https://www.youtube.com/watch?v=${item.snippet.resourceId.videoId}`,
            image_url: item.snippet.thumbnails.standard !== undefined ? item.snippet.thumbnails.standard.url : item.snippet.thumbnails.default.url,
            author: {
                id: channel.channel_id,
                name: channel.username,
                image_url: channel.author_image_url,
            },
        }
    }
}