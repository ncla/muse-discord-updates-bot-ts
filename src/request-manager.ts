import {WebhookMessageCreateOptions} from "discord.js";
import {RateLimiterMemory, RateLimiterQueue} from "rate-limiter-flexible";
import {retryPromise} from "@/src/common";
import {PromiseResult} from "@/src/types/promises";

export interface WebhookExecuteRequester<UpdateRequestBody> {
    send(update: UpdateRequestBody): Promise<Response>
    sendAll(): void
}

export abstract class AbstractUpdateRequestManager<UpdateRequestBody> implements WebhookExecuteRequester<UpdateRequestBody> {
    protected requestBodies: UpdateRequestBody[] = [];

    add(requestBody: UpdateRequestBody)
    {
        this.requestBodies.push(requestBody);
    }

    count(): number
    {
        return this.requestBodies.length;
    }

    abstract send(update: UpdateRequestBody): Promise<Response>;

    abstract sendAll(): void;
}

abstract class FixedWindowRateLimitedRequestManager<UpdateRequestBody> extends AbstractUpdateRequestManager<UpdateRequestBody> {
    async sendAll(requestAmountPerDuration: number = 5, rateLimitDurationSeconds: number = 2): Promise<PromiseResult<Response>[]>
    {
        const limiterFlexible = new RateLimiterMemory({
            points: requestAmountPerDuration,
            duration: rateLimitDurationSeconds,
        });

        const limiterQueue = new RateLimiterQueue(limiterFlexible);

        let sendPromises: Promise<PromiseResult<Response>>[] = []

        for (let i = 0; i < this.requestBodies.length; i++) {
            sendPromises.push(
                new Promise((resolve, reject) => {
                    limiterQueue
                        .removeTokens(1)
                        .then(() => {
                            retryPromise(
                                () => this.send(this.requestBodies[i]),
                                3,
                                2500
                            )
                            .then(result => resolve({status: 'fulfilled', value: result}))
                            .catch(error => resolve({status: 'rejected', reason: error}));
                        })
                        .catch((error) => {
                            resolve({status: 'rejected', reason: error})
                        })
                })
            )
        }

        return await Promise.all(sendPromises);
    }
}

export class DiscordWebhookRequestManager extends FixedWindowRateLimitedRequestManager<WebhookMessageCreateOptions> {
    private _webhookId: string;

    private _webhookToken: string;

    constructor(webhookId: string, webhookToken: string) {
        super();
        this._webhookId = webhookId;
        this._webhookToken = webhookToken;
    }

    async send(body: WebhookMessageCreateOptions): Promise<Response>
    {
        return await fetch(`https://discord.com/api/webhooks/${this._webhookId}/${this._webhookToken}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(body),
        })
        .then(async response => {
            if (!response.ok) {
                throw new Error(`Error: ${response.statusText}`);
            }

            return response;
        })
    }
}