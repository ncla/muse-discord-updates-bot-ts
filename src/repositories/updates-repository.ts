import { Kysely } from "kysely";
import {Database, InsertableUpdateRecord, SelectableUpdateRecord} from "@/src/database";

export interface IUpdatesRepository<CreateUpdateRecordType, ReturnableUpdateRecordType> {
    findByTypeAndUniqueId(type: string, uniqueId: string): Promise<ReturnableUpdateRecordType | undefined>;
    create(update: CreateUpdateRecordType): Promise<void> | void;
}

export class UpdatesRepositoryKysely implements IUpdatesRepository<InsertableUpdateRecord, SelectableUpdateRecord>
{
    constructor(private db: Kysely<Database>)
    {
        return this
    }

    async findByTypeAndUniqueId(type: string, uniqueId: string): Promise<SelectableUpdateRecord | undefined>
    {
        const result = await this
            .db
            .selectFrom('updates')
            .where('type', '=', type)
            .where('unique_id', '=', uniqueId)
            .selectAll()
            .executeTakeFirst()

        if (result && typeof result.data === 'string') {
            result.data = JSON.parse(result.data)
        }

        return result
    }

    async create(newUpdateRecord: InsertableUpdateRecord): Promise<void>
    {
        if (typeof newUpdateRecord.data === 'object') {
            newUpdateRecord.data = JSON.stringify(newUpdateRecord.data)
        }

        await this
            .db
            .insertInto('updates')
            .values(newUpdateRecord)
            .execute()
    }
}