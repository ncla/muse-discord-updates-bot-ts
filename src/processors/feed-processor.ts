import {EntryFetcher} from "@/src/entry-fetchers";
import {IUpdatesRepository} from "@/src/repositories/updates-repository";
import {getTransformer} from "@/src/updates/transformers";
import {
    WebhookExecuteRequestor
} from "@/src/webhook-requestor";
import {FixedWindowRateLimitedActionableQueueManager} from "@/src/action-queue-manager";
import {BaseUpdate, WebhookService, WebhookServiceBodyMap, WebhookServiceResponseMap} from "@/src/updates";
import {PromiseResult} from "@/src/types/promises";
import {FeedProcessorSummary, FetcherSummary, WebhookRequestSummary} from "../types/feed-processor";

export class FeedProcessor<
    CreateUpdateRecordType,
    ReturnableUpdateRecordType,
    WS extends WebhookService
>
{
    constructor(
        protected webhookService: WS,
        protected entryFetchers: EntryFetcher[],
        protected updatesRepository: IUpdatesRepository<CreateUpdateRecordType, ReturnableUpdateRecordType>,
        protected webhookExecuteRequestor: WebhookExecuteRequestor<WebhookServiceBodyMap[WS], WebhookServiceResponseMap[WS]>,
        protected queueActionManager: FixedWindowRateLimitedActionableQueueManager<WebhookServiceResponseMap[WS]>
    ) {
        return this
    }

    async process(): Promise<FeedProcessorSummary<WebhookServiceBodyMap[WS], WebhookServiceResponseMap[WS]>>
    {
        console.info('Running feed fetchers..')

        let requestPromises: Promise<PromiseResult<WebhookServiceResponseMap[WS]>>[] = []

        let fetcherSummaries: FetcherSummary<WebhookServiceBodyMap[WS]>[] = this.entryFetchers.map(fetcher => {
            return {
                name: fetcher.constructor.name,
                entries: [],
                entriesInDatabaseAlready: [],
                entriesProcessed: [],
                entriesTransformed: []
            }
        })

        let webhookRequestSummary: WebhookRequestSummary<WebhookServiceResponseMap[WS]> = {
            webhookService: this.webhookService,
            responses: []
        }

        await Promise.allSettled(
            this.entryFetchers.map(async (fetcher, fetcherIndex) => {
                const updates = await fetcher.fetch()

                console.log(`Fetched ${updates.length} updates from ${fetcher.constructor.name}`)

                fetcherSummaries[fetcherIndex].entries = updates

                return Promise.allSettled(
                    updates.map(async update => {
                        const transformedEntry = await this.processUpdateEntry(update)

                        if (transformedEntry === undefined) {
                            fetcherSummaries[fetcherIndex].entriesInDatabaseAlready.push(update)
                            return
                        }

                        fetcherSummaries[fetcherIndex].entriesProcessed.push(update)
                        fetcherSummaries[fetcherIndex].entriesTransformed.push(transformedEntry)

                        requestPromises.push(
                            this.queueActionManager.queue(
                                () => this.webhookExecuteRequestor.send(transformedEntry)
                            ).then(response => {
                                webhookRequestSummary.responses.push(response)

                                return response
                            })
                        )
                    })
                )
            })
        )

        await Promise.all(requestPromises)

        return {
            fetcherSummaries,
            webhookRequestSummary
        }
    }

    private async processUpdateEntry(entry: BaseUpdate): Promise<WebhookServiceBodyMap[WS] | undefined>
    {
        console.info(`Processing update entry: ${entry.type} – ${entry.uniqueId}`)

        const existingEntry = await this.updatesRepository.findByTypeAndUniqueId(
            entry.type,
            entry.uniqueId
        )

        if (existingEntry !== undefined) {
            console.info(`Entry already exists, skipping: ${entry.type} – ${entry.uniqueId}`)
            return
        }

        const newEntry = <CreateUpdateRecordType>{
            type: entry.type,
            unique_id: entry.uniqueId,
            data: entry
        }

        console.info(`Creating new entry: ${entry.type} – ${entry.uniqueId}`)

        this.updatesRepository.create(newEntry)

        console.info(`Created entry in database: ${entry.type} – ${entry.uniqueId}`)

        const entryTransformer = getTransformer(this.webhookService, entry.type)
        return entryTransformer.transform(entry) as WebhookServiceBodyMap[WS]
    }
}