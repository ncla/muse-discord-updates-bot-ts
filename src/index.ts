import {DiscordWebhookRequestManager} from "@/src/request-manager";
import {YoutubeUploads} from "@/src/entry-fetchers/youtube-uploads";
import config from "@/src/config";
import {WebhookService} from "@/src/updates";
import {UpdatesRepositoryKysely} from "@/src/repositories/updates-repository";
import {YoutubePlaylistsKysely} from "@/src/repositories/youtube-playlists-repository";
import {db, InsertableYoutubePlaylistRecord, ReturnableYoutubePlaylistRecord} from "@/src/database";
import {FeedProcessor} from "@/src/processors/feed-processor";
import {YoutubePlaylistVideos} from "@/src/entry-fetchers/youtube-playlists";
import {Musebootlegs} from "@/src/entry-fetchers/musebootlegs";

(async () => {
    const discordWebhookId = config.webhooks.discord.id;
    const discordWebhookToken = config.webhooks.discord.token;

    if (discordWebhookId === undefined || discordWebhookToken === undefined) {
        throw new Error('Discord webhook ID or token is not set');
    }

    const feedProcessor = new FeedProcessor(
        WebhookService.Discord,
        [
            new YoutubeUploads(config),
            new YoutubePlaylistVideos(new YoutubePlaylistsKysely(db), config),
            new Musebootlegs(config)
        ],
        new UpdatesRepositoryKysely(db),
        new DiscordWebhookRequestManager(discordWebhookId, discordWebhookToken)
    )

    feedProcessor.process()
})().catch(error => {
    console.error('Main app function error occurred:', error);
});