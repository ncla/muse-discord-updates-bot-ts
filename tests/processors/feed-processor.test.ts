import {afterEach, beforeAll, beforeEach, expect, test, vi} from 'vitest'
import {FeedProcessor} from "../../src/processors/feed-processor";
import {Update, UpdateType, WebhookService} from "../../src/update";
import {UpdatesRepositoryKysely} from "../../src/repositories/updates-repository";
import {clearTestDatabase, createTestDatabase} from "../__utils__/database";
import {DiscordWebhookRequestManager} from "../../src/request-manager";
import {YoutubeUploads} from "../../src/entry-fetchers/youtube-uploads";
import {createTestUnprocessedEntry} from "../__utils__";
import {YoutubeUpload as YoutubeUploadsTransformer} from "../../src/updates/transformers/discord/youtube-upload";
import * as transformerExports from '../../src/updates/transformers/index'
import config from "../../src/config";

const DB_FILE_IDENTIFIER = 'feed-processor'

beforeEach(async () => {
    await clearTestDatabase(DB_FILE_IDENTIFIER)
})

afterEach(async () => {
    await clearTestDatabase(DB_FILE_IDENTIFIER)
})

test('it processes with one of the fetchers throwing error', async () => {
    const db = await createTestDatabase(DB_FILE_IDENTIFIER)
    const updatesRepository = new UpdatesRepositoryKysely(db)
    const requestManager = new DiscordWebhookRequestManager('fake', 'fake')

    const entryFetcherGood = new YoutubeUploads(config)
    const entryFetcherFailing = new YoutubeUploads(config)

    entryFetcherGood.fetch = vi.fn(async () => {
        return <Update[]>[
            createTestUnprocessedEntry(UpdateType.YOUTUBE_UPLOAD),
        ]
    })

    entryFetcherFailing.fetch = vi.fn(async () => {
        throw new Error()
    })

    const requestManagerSendSpy = vi
        .spyOn(requestManager, 'send')
        .mockImplementation(async () => {
            return new Response()
        })

    const feedProcessor = new FeedProcessor(
        WebhookService.Discord,
        [entryFetcherGood, entryFetcherFailing],
        updatesRepository,
        requestManager
    )

    expect(async () => {
        await feedProcessor.process()
        expect(requestManagerSendSpy).toHaveBeenCalledTimes(1)
    }).not.toThrow()
})

test('insert query is not run when entry already exists', async () => {
    const FAKE_UNIQUE_ID = 'fake_unique_id';

    const db = await createTestDatabase(DB_FILE_IDENTIFIER)

    await db
        .insertInto('updates')
        .values({
            type: UpdateType.YOUTUBE_UPLOAD,
            unique_id: FAKE_UNIQUE_ID
        })
        .executeTakeFirst()

    const updatesRepository = new UpdatesRepositoryKysely(db)
    const requestManager = new DiscordWebhookRequestManager('fake', 'fake')

    const entryFetcherGood = new YoutubeUploads(config)

    let fakeYoutubeUploadEntry = createTestUnprocessedEntry(UpdateType.YOUTUBE_UPLOAD)
    fakeYoutubeUploadEntry.uniqueId = FAKE_UNIQUE_ID

    entryFetcherGood.fetch = vi.fn(async () => {
        return <Update[]>[
            fakeYoutubeUploadEntry,
        ]
    })

    const requestManagerSendSpy = vi
        .spyOn(requestManager, 'send')
        .mockImplementation(async () => {
            return new Response()
        })

    const updatesRepositoryFindSpy = vi.spyOn(updatesRepository, 'findByTypeAndUniqueId')
    const updatesRepositoryCreateSpy = vi.spyOn(updatesRepository, 'create')

    const feedProcessor = new FeedProcessor(
        WebhookService.Discord,
        [entryFetcherGood],
        updatesRepository,
        requestManager
    )

    await feedProcessor.process()

    expect(updatesRepositoryFindSpy).toHaveBeenCalledTimes(1)
    expect(updatesRepositoryCreateSpy).toHaveBeenCalledTimes(0)

    const rows = await db
        .selectFrom('updates')
        .selectAll()
        .execute()

    expect(rows.length).toBe(1)
})

test('it processes fetched entries in a loop without one entry failing entire process', async () => {
    const db = await createTestDatabase(DB_FILE_IDENTIFIER)

    const updatesRepository = new UpdatesRepositoryKysely(db)
    const requestManager = new DiscordWebhookRequestManager('fake', 'fake')
    const entryFetcher = new YoutubeUploads(config)

    let getTransformerCallCount = 0;

    // This could be any other method that returns error, we just need to test that the feed process continues
    vi.spyOn(transformerExports, 'getTransformer').mockImplementation(() => {
        if (getTransformerCallCount === 2) {
            getTransformerCallCount++
            throw new Error('xD')
        }

        getTransformerCallCount++
        return new YoutubeUploadsTransformer
    })

    const entries: Update[] = []

    for (let i = 0; i < 5; i++) {
        let entry = createTestUnprocessedEntry(UpdateType.YOUTUBE_UPLOAD)
        entry.uniqueId = `unique_id_${i}`
        entries.push(entry)
    }

    entryFetcher.fetch = vi.fn(async () => {
        return entries
    })

    const requestManagerSendSpy = vi
        .spyOn(requestManager, 'send')
        .mockImplementation(async () => {
            return new Response()
        })

    const feedProcessor = new FeedProcessor(
        WebhookService.Discord,
        [entryFetcher],
        updatesRepository,
        requestManager
    )

    await feedProcessor.process()

    expect(requestManagerSendSpy).toHaveBeenCalledTimes(4)
})