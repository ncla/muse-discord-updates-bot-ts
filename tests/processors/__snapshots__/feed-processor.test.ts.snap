// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feed manager stops the queue worker interval 1`] = `
{
  "fetcherSummaries": [
    {
      "entries": [
        {
          "author": {
            "id": "test",
            "image_url": "https://google.com/image.jpg",
            "name": "muse",
          },
          "content": "content",
          "created_at": 2025-01-01T00:00:00.000Z,
          "id": "test",
          "image_url": "https://google.com/image.jpg",
          "parent_id": null,
          "parent_title": null,
          "title": "title",
          "type": "YOUTUBE_UPLOAD",
          "uniqueId": "test",
          "url": "https://youtube.com/watch?v=test",
        },
      ],
      "entriesInDatabaseAlready": [],
      "entriesProcessed": [
        {
          "author": {
            "id": "test",
            "image_url": "https://google.com/image.jpg",
            "name": "muse",
          },
          "content": "content",
          "created_at": 2025-01-01T00:00:00.000Z,
          "id": "test",
          "image_url": "https://google.com/image.jpg",
          "parent_id": null,
          "parent_title": null,
          "title": "title",
          "type": "YOUTUBE_UPLOAD",
          "uniqueId": "test",
          "url": "https://youtube.com/watch?v=test",
        },
      ],
      "entriesTransformed": [
        {
          "content": "<@&700712850603245639> **muse** uploaded a video on YouTube",
          "embeds": [
            {
              "author": {
                "icon_url": "https://google.com/image.jpg",
                "name": "muse",
              },
              "color": 15158332,
              "description": "content",
              "footer": {
                "text": "Uploaded at 2025-01-01 00:00:00 UTC+0",
              },
              "thumbnail": {
                "url": "https://google.com/image.jpg",
              },
              "title": "title",
              "url": "https://youtube.com/watch?v=test",
            },
          ],
        },
      ],
      "errors": [],
      "name": "YoutubeUploads",
    },
  ],
  "webhookRequestSummary": {
    "errors": [],
    "responses": [
      Response {
        Symbol(realm): {
          "settingsObject": {},
        },
        Symbol(state): {
          "aborted": false,
          "cacheState": "",
          "headersList": HeadersList {
            "cookies": null,
            Symbol(headers map): Map {},
            Symbol(headers map sorted): null,
          },
          "rangeRequested": false,
          "requestIncludesCredentials": false,
          "status": 200,
          "statusText": "",
          "timingAllowPassed": false,
          "timingInfo": null,
          "type": "default",
          "urlList": [],
        },
        Symbol(headers): Headers {
          Symbol(guard): "response",
          Symbol(headers list): HeadersList {
            "cookies": null,
            Symbol(headers map): Map {},
            Symbol(headers map sorted): null,
          },
          Symbol(realm): {
            "settingsObject": {},
          },
        },
      },
    ],
    "webhookService": 0,
  },
}
`;
